### PR TITLE
Add trace support for IDONTWANT control messages

### PR DESCRIPTION
**Pull request by cortze**
# Decription
After the recent addition of `IDONTWANT` messages to the `go-libp2p-pubsub` repo ([GossipSub `v1.2`](https://github.com/libp2p/go-libp2p-pubsub/pull/553)), this PR aims to extend the libp2p/gossipsub tracer to track these same messages.

Some extra features:
- Update the `go-libp2p` and the `go-libp2p-pubsub` dependencies.
- Update the interface to define the `data-streamer` with a new flag a the [ `app-root` level](https://github.com/probe-lab/hermes/blob/d21bdb45edf86ae4e37ed5ad4cc8208fbeb0d7c8/cmd/hermes/cmd.go#L175). 
   - This should make local testing much simpler, along with the already merged PR:
       - #31 
- Extension of the Makefile to `build`, `install`, and `format` the code

### CODE REVIEW

**Comment by cortze** [comment]
I've been testing locally the changes (using the `logger` data-stream), and looks like it's working:
```
{"Type":"RECV_RPC","PeerID":"16Uiu2HAmGXD7jgbr6JRaAEP3sueEtgnmKt5Kz9CBsDwoaxh9M1iP","Timestamp":"2024-09-24T17:08:14.647175206+02:00","Data":{"PeerID":"16Uiu2HAmNyBDzhiZKGMRDLRbsbxwAyr6xDcBSPJwt4qazXrRYGss","Control":{"Idontwant":[{"MsgIDs":["c94a99f24e6ae9d3fc287592ebc7288ac77fec8e"]}]}}}
{"Type":"RECV_RPC","PeerID":"16Uiu2HAmGXD7jgbr6JRaAEP3sueEtgnmKt5Kz9CBsDwoaxh9M1iP","Timestamp":"2024-09-24T17:08:14.684288724+02:00","Data":{"PeerID":"16Uiu2HAmNyBDzhiZKGMRDLRbsbxwAyr6xDcBSPJwt4qazXrRYGss","Control":{"Idontwant":[{"MsgIDs":["b6a2851eac225ada7c7cc935ef480c57456c017e"]}]}}}
{"Type":"RECV_RPC","PeerID":"16Uiu2HAmGXD7jgbr6JRaAEP3sueEtgnmKt5Kz9CBsDwoaxh9M1iP","Timestamp":"2024-09-24T17:08:14.686044398+02:00","Data":{"PeerID":"16Uiu2HAmNyBDzhiZKGMRDLRbsbxwAyr6xDcBSPJwt4qazXrRYGss","Control":{"Idontwant":[{"MsgIDs":["b6a2851eac225ada7c7cc935ef480c57456c017e"]}]}}}
{"Type":"RECV_RPC","PeerID":"16Uiu2HAmGXD7jgbr6JRaAEP3sueEtgnmKt5Kz9CBsDwoaxh9M1iP","Timestamp":"2024-09-24T17:08:14.693322474+02:00","Data":{"PeerID":"16Uiu2HAmNyBDzhiZKGMRDLRbsbxwAyr6xDcBSPJwt4qazXrRYGss","Control":{"Idontwant":[{"MsgIDs":["e1e130e9d2487bb2d9f97f3497c63f6bd7bb68de"]}]}}}
{"Type":"RECV_RPC","PeerID":"16Uiu2HAmGXD7jgbr6JRaAEP3sueEtgnmKt5Kz9CBsDwoaxh9M1iP","Timestamp":"2024-09-24T17:08:14.696461042+02:00","Data":{"PeerID":"16Uiu2HAmNyBDzhiZKGMRDLRbsbxwAyr6xDcBSPJwt4qazXrRYGss","Control":{"Idontwant":[{"MsgIDs":["3e5f313168ed113b863a9a27b8827c730e4db06b"]}]}}}
{"Type":"RECV_RPC","PeerID":"16Uiu2HAmGXD7jgbr6JRaAEP3sueEtgnmKt5Kz9CBsDwoaxh9M1iP","Timestamp":"2024-09-24T17:08:14.700000251+02:00","Data":{"PeerID":"16Uiu2HAmNyBDzhiZKGMRDLRbsbxwAyr6xDcBSPJwt4qazXrRYGss","Control":{"Idontwant":[{"MsgIDs":["e1e130e9d2487bb2d9f97f3497c63f6bd7bb68de"]}]}}}
```
The only thing that worries me is that there seems to be a specific Control RPC for each of the IDONTWANT messages (I need to check if this is an intended behaviour)

File: go.mod [hunk 1]
```diff
@@ -6,19 +6,19 @@ require (
...
-	github.com/decred/dcrd/dcrec/secp256k1/v4 v4.2.0
+	github.com/decred/dcrd/dcrec/secp256k1/v4 v4.3.0
 	github.com/dennis-tra/go-kinesis v0.0.0-20240326083914-7acf5f8dc24e
 	github.com/ethereum/go-ethereum v1.13.15
 	github.com/ferranbt/fastssz v0.1.3
 	github.com/google/uuid v1.6.0
 	github.com/hashicorp/golang-lru/v2 v2.0.7
 	github.com/iand/pontium v0.3.14
-	github.com/libp2p/go-libp2p v0.33.1
+	github.com/libp2p/go-libp2p v0.36.2
```
[the remark this answers is not in the document]
Comment (cortze) [reply go.mod:16]: ohhh, it seems that the last release fixing https://github.com/libp2p/go-libp2p/commit/cd88f9784044da5466534713a23d61d874774a53 was launched after pushing the changes. 
`v0.36.3` was crashing at compilation time. Fixing it

File: host/host.go [hunk 2]
```diff
@@ -93,7 +93,7 @@ func (h *Host) Serve(ctx context.Context) error {
 				AgentVersion: h.AgentVersion(c.RemotePeer()),
 				Direction:    c.Stat().Direction.String(),
 				Opened:       c.Stat().Opened,
-				Transient:    c.Stat().Transient,
+				// Transient:    c.Stat().Transient,
```
[the remark this answers is not in the document]
Comment (guillaumemichel) [reply host/host.go:96]: IIRC `.Transient` has been replaced by `.Limited`

[the remark this answers is not in the document]
Comment (cortze) [reply host/host.go:96]: noted, I'll update it 👍🏽 thanks for the pointer Gui

File: eth/node_config.go [hunk 3]
```diff
@@ -76,10 +76,11 @@ type NodeConfig struct {
 	PrysmPortHTTP    int
 	PrysmPortGRPC    int
 
-	// The AWS Kinesis Data Stream configuration
-	AWSConfig     *aws.Config // if set, we consider Kinesis to be enabled
-	KinesisRegion string
-	KinesisStream string
+	// The Data Stream configuration
+	DataStreamType host.DataStreamType
+	AWSConfig      *aws.Config // if set, we consider Kinesis to be enabled
```
[the remark this answers is not in the document]
Comment (cortze) [reply eth/node_config.go]: addressed in commit https://github.com/probe-lab/hermes/pull/37/commits/c12a5397f91689264530143ca53c8c96d9da71e5

File: go.mod (hunk 1 above)
[the remark this answers is not in the document]
Comment (cortze) [reply go.mod]: addressed in commit https://github.com/probe-lab/hermes/pull/37/commits/c12a5397f91689264530143ca53c8c96d9da71e5

File: eth/node_config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cortze) [reply eth/node_config.go]: addressed in commit https://github.com/probe-lab/hermes/pull/37/commits/c12a5397f91689264530143ca53c8c96d9da71e5

File: host/trace_logger.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package host
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+)
+
+type TraceLogger struct{}
```
[the remark this answers is not in the document]
Comment (cortze) [reply host/trace_logger.go:9]: addressed in commit https://github.com/probe-lab/hermes/pull/37/commits/c12a5397f91689264530143ca53c8c96d9da71e5

**Review by guillaumemichel — APPROVED** [review]
LGTM

**Pull request merged**